src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,509 @@
+# - '_v-' of 'prefix_v-1.2.3'
+# - 'r_'  of 'r_3_3_3'
+# - 'c'   of 'c4.1'
+# - '.'   of 'name.9-9-9-9'
+PREFIX_SEPARATOR = "(?P<prefix_sep_0>(?:(?:(?<![0-9a-z])[vrc])|(?:[^0-9a-z][vrc])|[^0-9a-z])(?:[^0-9a-z])?)?"
+
+# Together, the prefix and prefix separator exist to separate the prefix from version part of a tag, while ensuring that
+# the prefix is free from non-prefix characters (the separator). Note that the prefix is expected to be at least two
+# characters in length to prevent overlap with separators and confusion with versions; the prefix separator is at most
+# three characters; and a negative lookback passes if there are no preceding characters.
+
+# The infix accepts either:
+# - One to three alphabetic characters.
+# - One to three non-alphanumeric characters.
+# Note: The upper limit of three could be reduced to two based on current data.
+INFIX_3 = "([a-z]{1,3}|[^0-9a-z]{1,3})"
+INFIX_1 = f"(?P<sep>{INFIX_3})"  # A named capture group of INFIX_3.
+INFIX_2 = "(?P=sep)"  # A back reference to INFIX_1.
+
+# The infix exists between parts of the version string. The most recent design resulted in use of a back reference to
+# ensure non-suffix version parts were separated by the same separator, e.g. 1.2.3 but not 1.2-3. However, one edge
+# case required this to be partially reverted, requiring 1.2-3 to be accepted, while another edge case where
+# additional zeros need to be added after a version to pad its length, e.g. 1.2 becomes 1.2.0.0, still requires it.
+
+# The suffix separator exists for much the same purpose as the prefix separator: splitting the suffix into the actual
+# suffix, and the characters that join it to the version.
+# It optionally accepts:
+# One to two non-alphanumeric characters that are followed by either:
+# - A non-numeric character (positive lookahead).
+# - No character of any kind (negative lookahead).
+# E.g.
+# - '_'  of 'prefix_1.2.3_suffix'
+# - '..  of 'name-v-4-4-4..RELEASE'
+# - '#'  of 'v0.0.1#'
+SUFFIX_SEPARATOR = "(?P<suffix_sep>(?:[^0-9a-z]{1,2}(?:(?=[^0-9])|(?!.))))?"
+
+# The suffix optionally accepts:
+# A string that starts with an alphabetic character, and continues for one or more characters of any kind.
+SUFFIX = "(?P<suffix>[a-z].*)?"
+
+# If a version string has less parts than this number it will be padded with additional zeros to provide better matching
+# opportunities.
+# For this to be applied, the version string must not have any non-numeric parts.
+# E.g 1.2 (2) -> 1.2.0.0 (4), 1.2.RELEASE (3) -> 1.2.RELEASE (3), 1.DEV-5 (3) -> 1.DEV-5 (3)
+MAX_ZERO_DIGIT_EXTENSION = 4
+
+split_pattern = re.compile("[^0-9a-z]", flags=re.IGNORECASE)
+validation_pattern = re.compile("[0-9a-z]+$", flags=re.IGNORECASE)
+alphabetic_only_pattern = re.compile("[a-z]+$", flags=re.IGNORECASE)
+numeric_only_pattern = re.compile("[0-9]+$")
+versioned_string = re.compile("[a-z]+[0-9]+$", flags=re.IGNORECASE)  # e.g. RC1, M5, etc.
+
+
+def find_commit(git_obj: Git, purl: PackageURL) -> tuple[str, str]:
```

Comment:
Can you please add a unit test for this function. If possible, try to pass invalid inputs that trigger error-handling branches.